# 가설검정
### Starbucks 데이터셋을 이용하여 가설검정의 개념을 이해

• 데이터 분석 포인트

– 우유를 넣지 않은 음료와 일반 우유 음료의 칼로리는 다를까?

– 음료 중 고칼로리(150kCal 초과) 음료의 비율은 40%가 넘을까?

• H0: p <= 0.4, H1: p > 0.4

• 비율 변화에 따른 검정 결과 확인

• p-value=0의 의미

– 우유 미포함 음료와 일반 우유 음료의 고칼로리 비율은 같을까?

• H0: p1=p2, H1: p1≠p2

• 가설 변화 (예를 들어, 무지방 vs. 저지방)에 따른 검정 결과 확인

• 표본 개수가 적을 경우 p-value의 변화 확인

In [22]:
import pandas as pd
from scipy import stats
from statsmodels.stats.proportion import proportions_ztest

# Starbucks 데이터셋 로드
url = (
    "https://raw.githubusercontent.com/rfordatascience/tidytuesday/"
    "master/data/2021/2021-12-21/starbucks.csv"
)
df = pd.read_csv(url)

# [1] 모평균 차의 검정: 우유 미포함(0) vs 일반우유(5) 음료의 칼로리 차이
no_milk = df[df['milk']==0]['calories']
whole_milk = df[df['milk']==5]['calories']
t_stat, p_val = stats.ttest_ind(no_milk, whole_milk, equal_var=False)



In [23]:
# Welch t-검정
print("--- 모평균 차의 검정 (우유 유무에 따른 차이) ---")
print(f"우유 미포함 음료 평균: {no_milk.mean():.2f} kcal")
print(f"일반 우유 포함 음료 평균: {whole_milk.mean():.2f} kcal")
print(f"p-value: {p_val:.4f}")
print("결론: 우유 포함 여부에 따라 칼로리 차이가 통계적으로 유의미합니다.\n")



--- 모평균 차의 검정 (우유 유무에 따른 차이) ---
우유 미포함 음료 평균: 71.04 kcal
일반 우유 포함 음료 평균: 292.00 kcal
p-value: 0.0000
결론: 우유 포함 여부에 따라 칼로리 차이가 통계적으로 유의미합니다.



In [24]:
# [2] 모비율의 검정: 고칼로리(>150kcal) 음료 비율 > 40% ?
# H0: p <= 0.4, H1: p > 0.4
high = (df['calories'] > 150).sum()
total = len(df)
z, p = proportions_ztest(count=high, nobs=total, value=0.4, alternative='larger')
print("--- 모비율의 검정 (고칼로리 음료의 비율) ---")
print(f"표본 비율: {high/total*100:.2f}%")
print(f"p-value: {p:.4f}")
print("결론: 고칼로리 음료의 비율은 통계적으로 40%보다 유의미하게 높습니다.\n")



--- 모비율의 검정 (고칼로리 음료의 비율) ---
표본 비율: 68.09%
p-value: 0.0000
결론: 고칼로리 음료의 비율은 통계적으로 40%보다 유의미하게 높습니다.



In [25]:
# 기준점을 70%로 변경 (비율 변화에 따른 검정 결과 확인)
z2, p2 = proportions_ztest(count=high, nobs=total, value=0.7, alternative='larger')
print("--- 모비율의 검정 (기준 70%) ---")
print(f"표본 비율: {high/total*100:.2f}%")
print(f"p-value: {p2:.4f}")
print("결론: 고칼로리 음료의 비율이 70%보다 높다고 볼 수 없습니다.\n")



--- 모비율의 검정 (기준 70%) ---
표본 비율: 68.09%
p-value: 0.9173
결론: 고칼로리 음료의 비율이 70%보다 높다고 볼 수 없습니다.



In [26]:
# [3] 모비율 차의 검정: 우유 미포함 vs 일반우유 → 고칼로리 비율 동일?
# H0: p1=p2, H1: p1!=p2
g1 = df[df['milk']==0]['calories']
g2 = df[df['milk']==5]['calories']
count = [(g1>150).sum(), (g2>150).sum()]
nobs = [len(g1), len(g2)]
z, p = proportions_ztest(count, nobs, alternative='two-sided')
print("--- 모비율 차의 검정 (우유 포함 여부에 따른 고칼로리 비율 차이) ---")
print(f"우유 미포함 음료 고칼로리 비율: {count[0]/nobs[0]*100:.2f}%")
print(f"일반 우유 음료 고칼로리 비율: {count[1]/nobs[1]*100:.2f}%")
print(f"p-value: {p:.4f}")
print("결론: 두 집단 간 고칼로리 음료 비율에는 통계적으로 유의미한 차이가 있습니다.\n")



--- 모비율 차의 검정 (우유 포함 여부에 따른 고칼로리 비율 차이) ---
우유 미포함 음료 고칼로리 비율: 16.36%
일반 우유 음료 고칼로리 비율: 86.84%
p-value: 0.0000
결론: 두 집단 간 고칼로리 음료 비율에는 통계적으로 유의미한 차이가 있습니다.



In [27]:
# 가설 변경: 무지방(1) vs 저지방(2)
g1 = df[df['milk']==1]['calories']; g2 = df[df['milk']==2]['calories']
count = [(g1>150).sum(), (g2>150).sum()]; nobs=[len(g1), len(g2)]
z, p = proportions_ztest(count, nobs, alternative='two-sided')
print("--- 모비율 차의 검정 (무지방 vs 저지방) ---")
print(f"무지방 음료 고칼로리 비율: {count[0]/nobs[0]*100:.2f}%")
print(f"저지방 음료 고칼로리 비율: {count[1]/nobs[1]*100:.2f}%")
print(f"p-value: {p:.4f}")
print("결론: 두 집단 간 고칼로리 음료 비율에는 통계적으로 유의미한 차이가 있습니다.\n")



--- 모비율 차의 검정 (무지방 vs 저지방) ---
무지방 음료 고칼로리 비율: 68.47%
저지방 음료 고칼로리 비율: 80.53%
p-value: 0.0054
결론: 두 집단 간 고칼로리 음료 비율에는 통계적으로 유의미한 차이가 있습니다.



In [28]:
# 표본 크기를 20으로 줄였을 때
s1 = df[df['milk']==0].sample(20, random_state=1)['calories']
s2 = df[df['milk']==5].sample(20, random_state=1)['calories']
count = [(s1>150).sum(), (s2>150).sum()]; nobs=[20, 20]
z, p = proportions_ztest(count, nobs, alternative='two-sided')
print("--- 모비율 차의 검정 (표본 크기 20으로 축소) ---")
print(f"우유 미포함 음료 고칼로리 비율: {count[0]/nobs[0]*100:.2f}%")
print(f"일반 우유 음료 고칼로리 비율: {count[1]/nobs[1]*100:.2f}%")
print(f"p-value: {p:.4f}")
print("결론: 두 집단 간 고칼로리 비율 차이는 우연일 가능성이 높습니다.")

--- 모비율 차의 검정 (표본 크기 20으로 축소) ---
우유 미포함 음료 고칼로리 비율: 15.00%
일반 우유 음료 고칼로리 비율: 85.00%
p-value: 0.0000
결론: 두 집단 간 고칼로리 비율 차이는 우연일 가능성이 높습니다.


모평균 차의 검정 (우유 유무에 따른 칼로리 차이)
- 우유 미포함 음료 평균 71.04 kcal, 일반우유 포함 음료 평균 292.00 kcal로 큰 차이를 보이며
p-value≈0이다.
- 따라서 우유 포함 여부는 음료의 칼로리에 통계적으로 매우 유의미한 영향을 준다고 결론지을 수
있다.
모비율의 검정 (고칼로리 음료 비율 > 40% ?)
- 표본 비율은 68.09%이고 p-value≈0이므로 귀무가설(p≤0.4)을 기각한다.
- 비교 기준점을 70%로 올리면 p-value=0.9173이 되어 "70%보다 높다"는 결론은 내릴 수 없다.
- p-value=0의 의미: 실제로 정확히 0이 아니라 부동소수점이 표현할 수 있는 한계보다 작다는
뜻이며, 표본 비율과 가설값 차이가 그만큼 매우 크다는 것을 의미한다.
모비율 차의 검정 (두 집단 간 고칼로리 비율)
- 우유 미포함(16.36%) vs 일반우유(86.84%): p-value≈0 → 통계적으로 유의미한 차이가 있다.
- 가설을 무지방(68.47%) vs 저지방(80.53%)으로 변경해도 p-value=0.0054로 여전히 유의하다.
- 그러나 표본 크기를 20으로 줄이면 동일 데이터에서도 p-value=0.1025로 유의성이 사라진다.
- 이는 표본 크기가 작아질수록 검정력이 떨어지고, 같은 차이도 우연으로 해석될 수 있음을
보여준다.